docs/design-documents/platform/clocks/clocks.md

```diff
@@ -0,0 +1,381 @@
+
+The real-time clock, if available, provides calendar date and time information, at one-second resolution. The RTC can continue to run while the CPU does not.
+
+The C APIs in `rtc_api.h` read and write the time as a C `time_t`.
+
+The class `RealTimeClock` provides a wrapper that acts as a Chrono clock, so that `RealTimeClock::time_point`can be used to express absolute real time.
+
+The methods `to_time_t` and `from_time_t` convert to and from C `time_t`.
+
+At present there are no other APIs using `RealTimeClock`, and no real-time-based wake or alarm facility, so utility is limited, but the class has been created for framework symmetry with the other clocks.
+
+### Comparison to standard C++ Chrono clocks
+
+Mbed OS does not provide the standard C++11 Chrono Clocks, as it is not possible to provide consistent behaviour across the various toolchains and libraries supported. Here's a suggested mapping from the 3 standard C++ Clocks to Mbed OS.
+
+**std::steady_clock**
+
+`std::steady_clock` represents a clock for which values of `time_point` never decrease as physical time advances and for which values of `time_point` advance at a steady rate relative to real time.
+
+`Kernel::Clock` can be used in place of `std::steady_clock` for general use.
+
+For slightly higher timestamp resolution you can use `LowPowerClock`, but if attempting to schedule time events, using RTOS APIs with `Kernel::Clock` will often work better due to the RTOS's deep sleep wake latency compensation.
+
+If high resolution is required, you can use `SteadyHighResClock`, but that will prevent deep sleep from occurring at all.
+
+**std::high_resolution_clock**
+
+`std::high_resolution_clock` represents the clock with the shortest tick period.
+
+`HighResClock` is the best choice for a high-resolution clock, but it will not run during deep sleep. You can use `HighResClock::lock` to keep it running, or otherwise ensure that deep sleep is locked.
+
+If not prepared to manually handle the locking, you can use `SteadyHighResClock`, which totally locks deep sleep.
+
+**std::system_clock**
+
+`std::system_clock` represents wall clock time from the system-wide realtime clock.
+
+`RealTimeClock` can be used in place of `std::system_clock`, and provides the same `to_time_t` and `from_time_t` methods. However, unlike many implementations, its resolution is only seconds.
+
+# Usage scenarios and examples
+
+### Using the RTOS clock
+
+**Basic use of Chrono time literals**
+
+The chrono namespace must be specified in a using declaration to activate the C++14 literals for durations.
+
+```C++
+#include "mbed.h"
+using namespace std::chrono_literals; // or std::chrono for all chrono types
+
+for (;;) {
+    LED1 = !LED1;
+    ThisThread::sleep_for(500ms);
+}
+```
+
+Following examples will not bother to show the `using` declaration.
+
+**Prefer absolute time**
+
+Using absolute time aids precision, guaranteeing that there is no drift due to code execution time. This applies to all clocks.
```

Comment:
To some extent the issue can be totally self avoiding, _if you're fast enough_. When you wake up at 500ms, if you complete your work before 501ms, then the granularity of timer ticks means `sleep_for(500ms)` will wake at 1000ms, then 1500ms, etc. You can't drift by less than a tick. 

Older documentation didn't cover this - this absolute time thing is a bugbear of mine, and I've been progressively adding more and more absolute APIs since 5.0.

This latest version of the PR adds `attach_absolute` to `Timeout` for the first time - that weren't there despite the underlying `TimerEvent` and `ticker_data_t` being based on absolute time since 5.1 or so.

Reply:
>  When you wake up at 500ms, if you complete your work before 501ms, then the granularity of RTOS timer ticks means sleep_for(500ms) will wake at 1000ms, then 1500ms, etc.

in my test I was printing to the serial port. I'll do more testing when this is merged :)

Reply:
As it so happens, my updated `Timer` example was measuring a print time. You could use that.

Using buffered serial output would help reduce the speed impact of logging. Turn on ` platform.stdio-buffered-serial` if you haven't already.

Reply:
Will do! Thanks!

Reply:
I've checked my experiment again. Here is the basic setup:

- 3 LEDs
- 3 threads
- each thread blinks a specific LED for every xx ms: 500ms, 1000ms and 1500ms respectively

I think the issue doesn't not come from logging but from the fact that events happen at the same time. After around 30 minutes (maybe less but I did not notice before), the blinking is out of sync.

Is there a way to prevent that?

Reply:
> Is there a way to prevent that?

Well, yes, use absolute time.  You'll only get away with the relative time being "okay if it's fast enough" if there's nothing else contending for resource in the system.

For threads, you could make those threads higher priority, so they at least get to pre-empt any boring other stuff. Otherwise you're stuck waiting for equal priority things to be pre-empted. (And will always defer to higher-priority things).

If you use `Ticker` to do it from interrupt instead of threads, then that's even higher priority, and it will use absolute time additions internally to keep you going at constant period. (But there would still be scope for start-up stagger, as the first call is on a relative time).

Also `EventQueue:call_in` uses absolute addition, IIRC.

Mind you, in a zero-load system, if the only 3 threads doing anything are blinking, I'd be a bit surprised to see them get out of sync. But any other thread printing could cause disruption.

Reply:
Note that absolute time is already available without this PR - you can do

    uint64_t t = Kernel::Clock::now();
    for (;;) {
        t += 500;
        ThisThread::sleep_until(t);
    }

Reply:
Fantastic! I'll try that out :)